### PR TITLE
bpo-24766: doc= argument to subclasses of property not handled correctly

### PR DESCRIPTION
**Pull request by embray**
Here's a PR for a small fix I've had languishing for it looks like a couple years now.  Do I need to add a blurb as well?

https://bugs.python.org/issue24766

#68954

### CODE REVIEW

**Review by ethanfurman — APPROVED** [review]

**Comment by mscuthbert** [comment]
should this have "awaiting merge" as a tag?

**Comment by embray** [comment]
I don't know. I guess?

File: Objects/descrobject.c [hunk 1]
```diff
@@ -1524,32 +1526,30 @@ property_init_impl(propertyobject *self, PyObject *fget, PyObject *fset,
...
-                    return -1;
-            }
+            Py_XSETREF(self->prop_doc, get_doc);
             self->getter_doc = 1;
         }
         else if (PyErr_ExceptionMatches(PyExc_Exception)) {
             PyErr_Clear();
         }
         else {
+            Py_DECREF(get_doc);
```
Comment (serhiy-storchaka) [inline Objects/descrobject.c]: `get_doc` is NULL here, isn't?

File: Objects/descrobject.c [hunk 2]
```diff
@@ -1524,32 +1526,30 @@ property_init_impl(propertyobject *self, PyObject *fget, PyObject *fset,
...
             return -1;
         }
     }
 
+    /* If this is a property subclass, put __doc__ in dict of the subclass
+       instance as well, otherwise it gets shadowed by __doc__ in the
+       class's dict. */
+    if (Py_TYPE(self) != &PyProperty_Type) {
+        int err = _PyObject_SetAttrId((PyObject *)self, &PyId___doc__,
+                                      self->prop_doc);
```
Comment (serhiy-storchaka) [inline Objects/descrobject.c]: `self->prop_doc` can be NULL.

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]

File: Objects/descrobject.c (hunk 1 above)
Comment (embray) [reply Objects/descrobject.c]: Yes, the decref is no good here.

**Comment by embray** [comment]
Forgot about this for a ah, long time.
Rebased on latest master and addressed Serhiy's review comments.

**Comment by serhiy-storchaka** [comment]
```
>>> class NoDoc:
...     @property
...     def __doc__(self):
...         raise AttributeError
... 
>>> class PropertySub(property):
...     """This is a subclass of property"""
... 
>>> property(NoDoc()).__doc__
>>> PropertySub(NoDoc()).__doc__
'This is a subclass of property'
```

**Comment by embray** [comment]
I see...  is there a test for this (bizarre) case?  Perhaps there should be?

**Comment by embray** [comment]
In fact, I'm not really sure what you're suggesting be done here.

**Comment by serhiy-storchaka** [comment]
I suggest that the `property` subclass should be consistent with `property`, i.e. return None.

There is no test for this case, and I suggest to add one.

**Comment by embray** [comment]
I agree, though this appears to be a different issue.  Your case is broken the same way with or without this patch and is not the same as the specific issue that this is fixing.  Though I don't mind fixing both as long as I'm monkeying around in there...

**Comment by serhiy-storchaka** [comment]
It looks just as a corner case of the original. I think it should be fixed in this PR.

**Comment by csabella** [comment]
@embray, please address Serhiy's last comment and also resolve the merge conflict.  Thank you!

**Comment by embray** [comment]
@csabella Rebased, and addressed @serhiy-storchaka's corner case, which I agree was related.

**Comment by embray** [comment]
Rebased again; @serhiy-storchaka's change request is implemented.

**Comment by embray** [comment]
Looks like it needs a minor rebase yet again.  But there's no other reason not to move forward on this minor bug fix, I think.

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by terryjreedy** [comment]
@embray Please click the button in  https://github.com/python/cpython/pull/2487#issuecomment-1857848446  to re-sign the CLA with your current GH email.  It is needed to merge here.

**Pull request merged**